pkg/assembler/backends/ent/backend/hasMetadata.go

```diff
@@ -0,0 +1,262 @@
+}
+
+func (b *EntBackend) IngestBulkHasMetadata(ctx context.Context, subjects model.PackageSourceOrArtifactInputs, pkgMatchType *model.MatchFlags, hasMetadataList []*model.HasMetadataInputSpec) ([]string, error) {
+	var results []string
+	for i := range hasMetadataList {
+		var subject model.PackageSourceOrArtifactInput
+		if len(subjects.Packages) > 0 {
+			subject = model.PackageSourceOrArtifactInput{Package: subjects.Packages[i]}
+		} else if len(subjects.Artifacts) > 0 {
+			subject = model.PackageSourceOrArtifactInput{Artifact: subjects.Artifacts[i]}
+		} else {
+			subject = model.PackageSourceOrArtifactInput{Source: subjects.Sources[i]}
+		}
+		hm, err := b.IngestHasMetadata(ctx, subject, pkgMatchType, *hasMetadataList[i])
+		if err != nil {
+			return nil, gqlerror.Errorf("IngestBulkHasMetadata failed with element #%v with err: %v", i, err)
+		}
+		results = append(results, hm.ID)
+	}
+	return results, nil
+}
+
+func hasMetadataPredicate(filter *model.HasMetadataSpec) predicate.HasMetadata {
+	predicates := []predicate.HasMetadata{
+		optionalPredicate(filter.ID, IDEQ),
+		optionalPredicate(filter.Since, hasmetadata.TimestampGTE),
+		optionalPredicate(filter.Key, hasmetadata.KeyEQ),
+		optionalPredicate(filter.Value, hasmetadata.ValueEQ),
+		optionalPredicate(filter.Justification, hasmetadata.JustificationEQ),
+		optionalPredicate(filter.Origin, hasmetadata.OriginEQ),
+		optionalPredicate(filter.Collector, hasmetadata.CollectorEQ),
+	}
+
+	if filter.Subject != nil {
+		switch {
+		case filter.Subject.Artifact != nil:
+			predicates = append(predicates, hasmetadata.HasArtifactWith(artifactQueryPredicates(filter.Subject.Artifact)))
+		case filter.Subject.Package != nil:
+			predicates = append(predicates, hasmetadata.Or(
+				hasmetadata.HasAllVersionsWith(packageNameQuery(pkgNameQueryFromPkgSpec(filter.Subject.Package))),
+				hasmetadata.HasPackageVersionWith(packageVersionQuery(filter.Subject.Package)),
+			))
+		case filter.Subject.Source != nil:
+			predicates = append(predicates, hasmetadata.HasSourceWith(sourceQuery(filter.Subject.Source)))
+		}
+	}
+	return hasmetadata.And(predicates...)
+}
+
+func upsertHasMetadata(ctx context.Context, client *ent.Tx, subject model.PackageSourceOrArtifactInput, pkgMatchType *model.MatchFlags, spec model.HasMetadataInputSpec) (*int, error) {
+	insert := client.HasMetadata.Create().
+		SetKey(spec.Key).
+		SetValue(spec.Value).
+		SetTimestamp(spec.Timestamp).
+		SetJustification(spec.Justification).
+		SetOrigin(spec.Origin).
+		SetCollector(spec.Collector)
+
+	conflictColumns := []string{
+		hasmetadata.FieldTimestamp,
+		hasmetadata.FieldKey,
+		hasmetadata.FieldValue,
+		hasmetadata.FieldJustification,
+		hasmetadata.FieldOrigin,
+		hasmetadata.FieldCollector,
+	}
+	var conflictWhere *sql.Predicate
+
+	switch {
+	case subject.Artifact != nil:
+		art, err := client.Artifact.Query().Where(artifactQueryInputPredicates(*subject.Artifact)).Only(ctx)
+		if err != nil {
+			return nil, err
```

Comment:
`fmt.Errorf` for a more descriptive error

Reply:
done